diff --git a/kernel-dts/t18x-common-modules/tegra186-camera-imx274-a00.dtsi b/kernel-dts/t18x-common-modules/tegra186-camera-imx274-a00.dtsi
index a987b45..32f6013 100644
--- a/kernel-dts/t18x-common-modules/tegra186-camera-imx274-a00.dtsi
+++ b/kernel-dts/t18x-common-modules/tegra186-camera-imx274-a00.dtsi
@@ -62,8 +62,8 @@
 	};
 
 	i2c@3180000 {
-		tca9546@70 {
-			i2c@0 {
+		//tca9546@70 {
+		//	i2c@0 {
 			imx274_a@1a {
 			compatible = "nvidia,imx274";
 			/* I2C device address */
@@ -76,7 +76,6 @@
 			physical_w = "3.674";
 			physical_h = "2.738";
 
-			sensor_model = "imx274";
 
 			/* Define any required hw resources needed by driver */
 			/* ie. clocks, io pins, power sources */
@@ -182,48 +181,99 @@
 					active_h = "2160";
 					pixel_t = "bayer_rggb";
 					readout_orientation = "90";
-					line_length = "4208";
+					line_length = "4200";
 					inherent_gain = "1";
-					mclk_multiplier = "24";
-					pix_clk_hz = "576000000";
+					mclk_multiplier = "25";
+					pix_clk_hz = "756000000";
 
 					min_gain_val = "1.0";
-					max_gain_val = "22.2";
+					max_gain_val = "180";
 					min_hdr_ratio = "1";
-					max_hdr_ratio = "1";
-					min_framerate = "1.5";
+					max_hdr_ratio = "64";
+					min_framerate = "1.462526";
 					max_framerate = "60";
-					min_exp_time = "44";
-					max_exp_time = "666637";
-					embedded_metadata_height = "1";
+					min_exp_time = "16.165";
+					max_exp_time = "165770";
+					//embedded_metadata_height = "1";
 				};
-				mode1 { // IMX274_MODE_1920X1080
-					mclk_khz = "24000";
-					num_lanes = "4";
-					tegra_sinterface = "serial_a";
-					discontinuous_clk = "yes";
-					dpcm_enable = "false";
-					cil_settletime = "0";
+                                mode1 { // IMX274_MODE_3840X2160
+                                        mclk_khz = "24000";
+                                        num_lanes = "4";
+                                        tegra_sinterface = "serial_a";
+                                        discontinuous_clk = "yes";
+                                        dpcm_enable = "false";
+                                        cil_settletime = "0";
 
-					active_w = "1920";
-					active_h = "1080";
-					pixel_t = "bayer_rggb";
-					readout_orientation = "90";
-					line_length = "4160";
-					inherent_gain = "1";
-					mclk_multiplier = "24";
-					pix_clk_hz = "576000000";
+                                        active_w = "1920";
+                                        active_h = "1080";
+                                        pixel_t = "bayer_rggb";
+                                        readout_orientation = "90";
+                                        line_length = "4200";
+                                        inherent_gain = "1";
+                                        mclk_multiplier = "25";
+                                        pix_clk_hz = "756000000";
 
-					min_gain_val = "1.0";
-					max_gain_val = "22.2";
-					min_hdr_ratio = "1";
-					max_hdr_ratio = "1";
-					min_framerate = "1.5";
-					max_framerate = "60";
-					min_exp_time = "58";
-					max_exp_time = "184611";
-					embedded_metadata_height = "1";
-				};
+                                        min_gain_val = "1.0";
+                                        max_gain_val = "180";
+                                        min_hdr_ratio = "1";
+                                        max_hdr_ratio = "64";
+                                        min_framerate = "1.462526";
+                                        max_framerate = "60";
+                                        min_exp_time = "16.165";
+                                        max_exp_time = "165770";
+                                };
+                                mode2 { // IMX274_MODE_1280X720
+                                        mclk_khz = "24000";
+                                        num_lanes = "4";
+                                        tegra_sinterface = "serial_a";
+                                        discontinuous_clk = "yes";
+                                        dpcm_enable = "false";
+                                        cil_settletime = "0";
+
+                                        active_w = "1280";
+                                        active_h = "720";
+                                        pixel_t = "bayer_rggb";
+                                        readout_orientation = "90";
+                                        line_length = "4200";
+                                        inherent_gain = "1";
+                                        mclk_multiplier = "25";
+                                        pix_clk_hz = "756000000";
+
+                                        min_gain_val = "1.0";
+                                        max_gain_val = "180";
+                                        min_hdr_ratio = "1";
+                                        max_hdr_ratio = "64";
+                                        min_framerate = "1.462526";
+                                        max_framerate = "60";
+                                        min_exp_time = "16.165";
+                                        max_exp_time = "165770";
+                                };
+                                mode3 { // IMX274_MODE_1280X540
+                                        mclk_khz = "24000";
+                                        num_lanes = "4";
+                                        tegra_sinterface = "serial_a";
+                                        discontinuous_clk = "yes";
+                                        dpcm_enable = "false";
+                                        cil_settletime = "0";
+
+                                        active_w = "1280";
+                                        active_h = "540";
+                                        pixel_t = "bayer_rggb";
+                                        readout_orientation = "90";
+                                        line_length = "4200";
+                                        inherent_gain = "1";
+                                        mclk_multiplier = "25";
+                                        pix_clk_hz = "756000000";
+
+                                        min_gain_val = "1.0";
+                                        max_gain_val = "180";
+                                        min_hdr_ratio = "1";
+                                        max_hdr_ratio = "64";
+                                        min_framerate = "1.462526";
+                                        max_framerate = "240";
+                                        min_exp_time = "16.165";
+                                        max_exp_time = "165770";
+                                };
 				ports {
 					#address-cells = <1>;
 					#size-cells = <0>;
@@ -237,8 +287,8 @@
 						};
 					};
 				};
-			};
-		};
+			//};
+		//};
 	};
 	lens_imx274@A6V26 {
 		min_focus_distance = "0.0";
@@ -302,21 +352,21 @@
 		modules {
 			module0 {
 				badge = "imx274_bottom_A6V26";
-				position = "rear";
+				position = "bottom";
 				orientation = "1";
 				drivernode0 {
 					/* Declare PCL support driver (classically known as guid)  */
 					pcl_id = "v4l2_sensor";
 					/* Driver v4l2 device name */
-					devname = "imx274 30-001a";
+					devname = "imx274 2-001a";
 					/* Declare the device-tree hierarchy to driver instance */
-					proc-device-tree = "/proc/device-tree/i2c@3180000/tca9546@70/i2c@0/imx274_a@1a";
-				};
-				drivernode1 {
-					/* Declare PCL support driver (classically known as guid)  */
-					pcl_id = "v4l2_lens";
-					proc-device-tree = "/proc/device-tree/lens_imx274@A6V26/";
+					proc-device-tree = "/proc/device-tree/i2c@3180000/imx274_a@1a";
 				};
+                                drivernode1 {
+                                        /* Declare PCL support driver (classically known as guid)  */
+                                        pcl_id = "v4l2_lens";
+                                        proc-device-tree = "/proc/device-tree/lens_imx274@A6V26/";
+                                };
 			};
 		};
 	};
diff --git a/kernel-dts/t18x-common-platforms/tegra186-quill-camera-imx274-a00.dtsi b/kernel-dts/t18x-common-platforms/tegra186-quill-camera-imx274-a00.dtsi
index a700273..042c14c 100644
--- a/kernel-dts/t18x-common-platforms/tegra186-quill-camera-imx274-a00.dtsi
+++ b/kernel-dts/t18x-common-platforms/tegra186-quill-camera-imx274-a00.dtsi
@@ -17,7 +17,10 @@
 #include <t18x-common-modules/tegra186-camera-imx274-a00.dtsi>
 #include "dt-bindings/clock/tegra186-clock.h"
 
+/*
 #define CAM0_RST_L	TEGRA_MAIN_GPIO(R, 5)
+*/
+#define CAM0_RST_L	TEGRA_MAIN_GPIO(T, 5)
 #define CAM0_PWDN	TEGRA_MAIN_GPIO(R, 0)
 #define CAM1_RST_L	TEGRA_MAIN_GPIO(R, 1)
 #define CAM1_PWDN	TEGRA_MAIN_GPIO(L, 6)
@@ -27,22 +30,6 @@
 
 / {
 	i2c@3180000 {
-		tca9546@70 {
-			compatible = "nxp,pca9546";
-			reg = <0x70>;
-			#address-cells = <1>;
-			#size-cells = <0>;
-			skip_mux_detect = "yes";
-			vif-supply = <&en_vdd_cam>;
-                        vcc-supply = <&en_vdd_cam>;
-			vcc_lp = "vif";
-                        force_bus_start = <CAMERA_I2C_MUX_BUS(0)>;
-
-			i2c@0 {
-				reg = <0>;
-				i2c-mux,deselect-on-exit;
-				#address-cells = <1>;
-				#size-cells = <0>;
 				pca9570_a@24 {
 					compatible = "nvidia,pca9570";
 					reg = <0x24>;
@@ -62,7 +49,5 @@
 					vif-supply = <&en_vdd_cam>;
 					vdig-supply = <&en_vdd_cam_1v2>;
 				};
-			};
-		};
-	};
+                        };
 };
diff --git a/kernel-dts/t18x-common-platforms/tegra186-quill-camera-li-mipi-adpt-a00.dtsi b/kernel-dts/t18x-common-platforms/tegra186-quill-camera-li-mipi-adpt-a00.dtsi
index d456bf0..63c3d0d 100644
--- a/kernel-dts/t18x-common-platforms/tegra186-quill-camera-li-mipi-adpt-a00.dtsi
+++ b/kernel-dts/t18x-common-platforms/tegra186-quill-camera-li-mipi-adpt-a00.dtsi
@@ -17,7 +17,7 @@
 #include <t18x-common-modules/tegra186-camera-li-mipi-adpt-a00.dtsi>
 #include "dt-bindings/clock/tegra186-clock.h"
 
-#define CAM0_RST_L	TEGRA_MAIN_GPIO(R, 5)
+#define CAM0_RST_L	TEGRA_MAIN_GPIO(T, 5)
 #define CAM0_PWDN	TEGRA_MAIN_GPIO(R, 0)
 #define CAM1_RST_L	TEGRA_MAIN_GPIO(R, 1)
 #define CAM1_PWDN	TEGRA_MAIN_GPIO(L, 6)
diff --git a/kernel-dts/t18x-common-platforms/tegra186-quill-camera-modules.dtsi b/kernel-dts/t18x-common-platforms/tegra186-quill-camera-modules.dtsi
index 859009d..8e6ede9 100644
--- a/kernel-dts/t18x-common-platforms/tegra186-quill-camera-modules.dtsi
+++ b/kernel-dts/t18x-common-platforms/tegra186-quill-camera-modules.dtsi
@@ -18,7 +18,7 @@
 #include "t18x-common-platforms/tegra186-quill-camera-e3322-a00.dtsi"
 #include "t18x-common-platforms/tegra186-quill-camera-li-mipi-adpt-a00.dtsi"
 #include "t18x-common-platforms/tegra186-quill-camera-imx274-a00.dtsi"
-#include "t18x-common-platforms/tegra186-quill-camera-vivid.dtsi"
+//#include "t18x-common-platforms/tegra186-quill-camera-vivid.dtsi"
 
 #define CAM0_RST_L	TEGRA_MAIN_GPIO(R, 5)
 #define CAM0_PWDN	TEGRA_MAIN_GPIO(R, 0)
